File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
    let dummy = new ListNode(0);
    dummy.next = head;
    let prev = dummy;

    while(prev.next !== null && prev.next.next !== null){
        // Nodes to swap
        let first = prev.next;
        let second = prev.next.next;

        // Swapping nodes
        prev.next = second;
        first.next = second.next;
        second.next = first;

        prev = first;
    }
    return dummy.next;
};

File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Create dummy node and point it to head
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        
        while prev.next and prev.next.next:
            # Identify nodes to be swapped
            first = prev.next
            second = prev.next.next
            # Perform swap
            prev.next = second
            first.next = second.next
            second.next = first
            prev = first
        return dummy.next


